### PR TITLE
513/handle null frequencies

### PR DESCRIPTION
**Pull request by bderusha**
Issue Number: 513

## Summary
Improve handling of null frequency values from WattTime

## Changes

- Use the distance between dataPoints as the default duration, rather than Zero
- New unit tests to reflect this behavior ([changes here](https://github.com/microsoft/carbon-aware-sdk/pull/117/files#diff-248efed1bc704787aa3472bf81b17048d1a99d755c9b73e107da03163071b99aR255-R288)).
- Refactor unit tests to reduce boilerplate.

## Checklist

- [x] Local Tests Passing?
- [x] CICD and Pipeline Tests Passing?
- [x] Added any new Tests? YES
- [x] Documentation Updates Made? NO
- [x] Are there any API Changes? NO
- [x] This is not a breaking change. If it is, please describe it below.

### CODE REVIEW

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:161]: 💯 much cleaner to put in here

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:162]: Just for my own knowledge, if the array is empty, the `.Skip(1)` call won't crash, just goes to null as if there was only 1 element as well?

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs]: Why do we do `.Duration()` now instead of how it was?

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:172]: nice!

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs:255]: I would make a note that the actual changes in this file (aside from refactors) actually start here.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs]: it feels wrong to potentially have an ArugmentNullException in a test file where we are intentionally passing in valid default locations.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs]: Shouldn't `endDate` be some derivative of the `start` field so that if the default `DataStartTime` ever changes, this test is still valid

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs
Comment (bderusha) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs:255]: Added a link in the PR description and will leave this comment open until we're ready to merge👍

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs:288]: Does `CollectionAssert` work if the elements aren't in the same order? Or do we always expect the test case to match the output order of the durations

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs
Comment (bderusha) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs]: This was a c+p of the preexisting `SetupBalancingAuthority` method.  I believe it is just for type safety.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (pritipath) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs]: `.Duration()` gives an absolute value of the given Timespan object and thats what we wanted

**Review by JenMadiedo — CHANGES_REQUESTED** [review]
Have some small Qs but mostly lgtm

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (bderusha) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:162]: Correct, but I will add a test case to ensure.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs
Comment (pritipath) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs:288]: CollectionAssert checks for the order of elements in the collection too

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (bderusha) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:167]: Because WattTime data can be chronological OR reverse chronological, `.Duration()` gives us the absolute value.

**Review by KanishkT123 — APPROVED** [review]
This LGTM! Shipit.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (gfmatthews) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs]: It might be more stylistically accurate and easier to read if you used the TimeSpan.Subtract method directly (and omitted the .Duration)

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (gfmatthews) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:162]: Might it be easier to just check in this call in the array is at least 2 elements long?

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs
Comment (bderusha) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs]: ### ⚠️ ALERT ⚠️ 
line-ending changes are making this diff go bananas 🍌 .   I am fixing it

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (juzuluag) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:162]: > Just for my own knowledge, if the array is empty, the `.Skip(1)` call won't crash, just goes to null as if there was only 1 element as well?

Same concern here. How about the 1st record?

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs
Comment (juzuluag) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs]: nit: Seems like `frequency` is not set anywhere. Suggestion is to keep it simple and not to pass args that are not used by anyone.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (bderusha) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:162]: The reason for this approach is that it precisely accesses only the datapoints required for calculation, whereas `Count` is going to iterate through the entire enumerable.

Given that this approach is a direct port of the current implementation, we will likely leave it as is unless someone feels strongly about the readability.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (bderusha) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs]: @gfmatthews Agreed on `TimeSpan.Subtract`.  But we still need to use `Duration` to handle getting the absolute value.  I've added a comment to that effect

Comment (pritipath) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs]: `Subtract` method could return a negative value and we would still need to use Duration() to convert it.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (bderusha) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:162]: @juzuluag @JenMadiedo Here are the test cases exercising 0, 1, and 2 element responses through this code path.  https://github.com/microsoft/carbon-aware-sdk/pull/117/files#diff-248efed1bc704787aa3472bf81b17048d1a99d755c9b73e107da03163071b99aR258-R262

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs
Comment (pritipath) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs]: frequency is used to set `Frequency` in the `GridEmissionDataPoint` object

**Review by JenMadiedo — APPROVED** [review]
🚢 -it

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs
Comment (bderusha) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs]: @pritipath in this case I think @juzuluag means that the argument is never used by any of the functions that call this method.  He is right and it is easy to add back if we write any unit tests that could benefit from passing this in.

I agree that as a matter of good practice we should not be writing "just-in-case" code, and instead wait until we need it.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (gfmatthews) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs]: Right but if we get a negative value out of subtract, that's an indication that something else has gone wrong isn't it?

Comment (bderusha) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs]: @gfmatthews In this case, no.  WattTime will return data in chronological order as well as reverse-chronological order but we only care about the distance between the points, not the direction.

Chronological: `2022-08-01T12:00:00Z` - `2022-08-01T12:05:00Z` = `- 00:05:00`  (negative)
Reverse-Chronological: `2022-08-01T12:05:00Z` - `2022-08-01T12:00:00Z` = `00:05:00` (positive)

Both are valid for estimating the positive `frequency` or `duration` of the individual data points.

**Review by dnastrain — APPROVED** [review]
Overall LGTM, :shipit: -- had a quick call with bderusha to get my additional questions answered, such as: were we fixing the chronology problem and found the defaulting one, or vice-versa.

**Pull request merged**
